node/core/backing/src/lib.rs

```diff
@@ -0,0 +1,1149 @@
+		let (tx_to, rx_to) = mpsc::channel(CHANNEL_CAPACITY);
+		let (tx_from, rx_from) = mpsc::channel(CHANNEL_CAPACITY);
+
+		(
+			Self {
+				keystore,
+				parent,
+				rx_to,
+				tx_to,
+				tx_from,
+
+				validation_code: ValidationCode::default(),
+				head_data: HeadData::default(),
+				assignment: ParaId::default(),
+
+				issued_validity: HashSet::default(),
+				backed_candidates: HashSet::default(),
+				seconded: None,
+				backing_watchers: Vec::new(),
+				reported_misbehavior_for: HashSet::default(),
+
+				table: Table::default(),
+				table_context: TableContext::default(),
+			},
+			rx_from,
+		)
+	}
+
+	/// Hand out the sending part of the channel to communicate with this job.
+	fn tx(&mut self) -> mpsc::Sender<ToJob> {
+		self.tx_to.clone()
+	}
+
+	/// Request a validator set from the `RuntimeApi`.
+	async fn request_validators(&mut self) -> Result<Vec<ValidatorId>, Error> {
+		let (tx, rx) = oneshot::channel();
+
+		self.tx_from.send(FromJob::RuntimeApiMessage(RuntimeApiMessage::Request(
+				self.parent.clone(),
+				RuntimeApiRequest::Validators(tx),
+			)
+		)).await?;
+
+		Ok(rx.await?)
+	}
+
+	/// Request the scheduler roster from `RuntimeApi`.
+	async fn request_validator_groups(&mut self) -> Result<SchedulerRoster, Error> {
+		let (tx, rx) = oneshot::channel();
+
+		self.tx_from.send(FromJob::RuntimeApiMessage(RuntimeApiMessage::Request(
+				self.parent.clone(),
+				RuntimeApiRequest::ValidatorGroups(tx),
+			)
+		)).await?;
+
+		Ok(rx.await?)
+	}
+
+	/// Request the validation code for some `ParaId` at given block
+	/// from `RuntimeApi`.
+	async fn request_validation_code(
+		&mut self,
+		id: ParaId,
+		parent: BlockNumber,
+		parablock: Option<BlockNumber>,
+	) -> Result<ValidationCode, Error> {
+		let (tx, rx) = oneshot::channel();
+
+		self.tx_from.send(FromJob::RuntimeApiMessage(RuntimeApiMessage::Request(
+				self.parent.clone(),
+				RuntimeApiRequest::ValidationCode(id, parent, parablock, tx),
+			)
+		)).await?;
+
+		Ok(rx.await?)
+	}
+
+	/// Request a `SigningContext` from the `RuntimeApi`.
+	async fn request_signing_context(&mut self) -> Result<SigningContext, Error> {
+		let (tx, rx) = oneshot::channel();
+
+		self.tx_from.send(FromJob::RuntimeApiMessage(RuntimeApiMessage::Request(
+				self.parent.clone(),
+				RuntimeApiRequest::SigningContext(tx),
+			)
+		)).await?;
+
+		Ok(rx.await?)
+	}
+
+	/// Request `HeadData` for some `ParaId` from `RuntimeApi`.
+	async fn request_head_data(&mut self, id: ParaId) -> Result<HeadData, Error> {
+		let (tx, rx) = oneshot::channel();
+
+		self.tx_from.send(FromJob::RuntimeApiMessage(RuntimeApiMessage::Request(
+				self.parent.clone(),
+				RuntimeApiRequest::HeadData(id, tx),
+			)
+		)).await?;
+
+		Ok(rx.await?)
+	}
+
+	/// Logic that runs on startup.
+	async fn on_startup(&mut self) -> Result<(), Error> {
+		let validators = self.request_validators().await?;
+	
+		let roster = self.request_validator_groups().await?; 
```

Comment:
maybe slightly more efficient to `join` these runtime API requests so they can be pipelined.

At some point I'd like to do a refactor where we return most of this info from a single runtime API.

Reply:
I don't think that it's doable: it would require two `&mut self` refs

Reply:
Yeah, this is one of the "Bad side-effects" of Object-Oriented design in Rust. In the functional style you can get away with this more easily because the lifetime of the return type of `request_validators` should not depend on the lifetime of the borrow of self.

Reply:
If I am understanding you right it would run into the same problem unless we do this by hand without any functions at all. Otherwise there still is a single channel we do the communication on and it would require passing two mutable references into two funcions:

```rust
async fn do_this(s: &mut mpsc::Sender<()>) { s.send(()).await; }

async fn do_that(s: &mut mpsc::Sender<()> { s.send(()).await; }

join(do_this(&mut tx), do_that(&mut tx));
```
Or clone the `Sender` and pass it by value.

Or do everything in place: send message1 and get the `oneshot::Receiver`, send message 2 and get the `oneshot::Receiver` and do a join on those two.

Reply:
> Or do everything in place: send message1 and get the oneshot::Receiver, send message 2 and get the oneshot::Receiver and do a join on those two.

Yeah, this is what I was thinking of. Because as written, we wait for the response on one channel before even sending the message on the other.

If `async fn request_x(&mut Sender) -> impl Future<Item = Result<X, SendError>>` works it might be the best way to implement that.